File: tests/test_matcher.py
import pickle

import pytest
from hypothesis import given, settings
from hypothesis import strategies as st

from ahocorasick_demo import PatternMatcher

# A small alphabet makes collisions, overlaps, and nested patterns common, which
# is exactly where an automaton is easy to get wrong.
small_alphabet = st.text(alphabet="abc", min_size=1, max_size=6)
pattern_lists = st.lists(small_alphabet, min_size=1, max_size=8, unique=True)
texts = st.text(alphabet="abc", max_size=60)


def reference_find_all(patterns, text):
    """Obviously-correct O(n*k) oracle, including overlaps."""
    found = []
    for start in range(len(text)):
        for index, pattern in enumerate(patterns):
            if text.startswith(pattern, start):
                found.append((start, index))
    return sorted(found)


# --- Core behaviour ----------------------------------------------------------
#
# The fixed-case core behaviour - overlaps, nesting, counting, state counts, the
# empty pattern list, repeated patterns - lives in tests/test_matcher.cpp. It
# needs no interpreter to state, and stating it in C++ means a failure points at
# the algorithm rather than arriving through the binding layer.
#
# What stays here is what genuinely exercises the whole stack: the binding
# layer's own behaviour, and the property tests against a brute-force oracle.


def test_rejects_empty_patterns():
    # An empty pattern matches everywhere, which is never what a caller means.
    # The C++ suite pins the exception *type*; this pins nanobind's translation
    # of it into ValueError, which is the part Python callers depend on.
    with pytest.raises(ValueError, match="empty"):
        PatternMatcher(["ok", ""])


def test_dunder_len_and_repr_are_wired_to_the_automaton():
    # These two have no C++ counterpart to move to: __len__ and __repr__ are
    # defined in bindings.cpp, so the mapping from num_patterns/num_states onto
    # Python's protocols is only observable from here.
    matcher = PatternMatcher(["he", "she", "hers"])

    assert len(matcher) == 3
    assert matcher.num_states == 8
    assert repr(matcher) == "PatternMatcher(patterns=3, states=8)"


# --- str / bytes and offsets -------------------------------------------------


def test_str_and_bytes_patterns_are_interchangeable():
    from_str = PatternMatcher(["café"])
    from_bytes = PatternMatcher(["café".encode()])

    assert from_str.patterns == from_bytes.patterns == [b"caf\xc3\xa9"]


def test_str_offsets_are_code_points_not_bytes():
    # "é" is two bytes in UTF-8. Reporting byte offsets for str input would
    # disagree with every offset Python itself produces.
    text = "café bar"
    matcher = PatternMatcher(["bar"])

    assert matcher.find_all(text) == [(5, 0)]
    assert text[5:8] == "bar"


def test_bytes_offsets_are_byte_positions():
    matcher = PatternMatcher([b"bar"])
    text = "café bar".encode()

    assert matcher.find_all(text) == [(6, 0)]
    assert text[6:9] == b"bar"


def test_patterns_are_returned_as_bytes_in_order():
    matcher = PatternMatcher(["b", "a"])

    assert matcher.patterns == [b"b", b"a"]


# --- Properties, against a brute-force oracle --------------------------------


@given(patterns=pattern_lists, text=texts)
@settings(max_examples=300)
def test_find_all_agrees_with_brute_force(patterns, text):
    matcher = PatternMatcher(patterns)

    assert sorted(matcher.find_all(text)) == reference_find_all(patterns, text)


@given(patterns=pattern_lists, text=texts)
def test_count_agrees_with_find_all(patterns, text):
    matcher = PatternMatcher(patterns)

    assert matcher.count(text) == len(matcher.find_all(text))


@given(patterns=pattern_lists, text=texts)
def test_matches_agrees_with_count(patterns, text):
    matcher = PatternMatcher(patterns)

    assert matcher.matches(text) == (matcher.count(text) > 0)


@given(patterns=pattern_lists, text=texts)
def test_str_and_bytes_scanning_find_the_same_occurrences(patterns, text):
    matcher = PatternMatcher(patterns)

    # ASCII-only here, so code point offsets and byte offsets coincide.
    assert matcher.find_all(text) == matcher.find_all(text.encode())


# --- Pickle ------------------------------------------------------------------


@given(patterns=pattern_lists, text=texts)
def test_pickle_round_trips(patterns, text):
    matcher = PatternMatcher(patterns)

    restored = pickle.loads(pickle.dumps(matcher))

    assert restored.patterns == matcher.patterns
    assert restored.num_states == matcher.num_states
    assert restored.find_all(text) == matcher.find_all(text)


def test_handles_patterns_that_are_not_valid_utf8():
    # Arbitrary binary is a legitimate pattern: the matcher works on bytes, and
    # nothing should try to decode them as text on the way in or out.
    matcher = PatternMatcher([b"\xff\xfe", b"\x00\x01"])

    assert matcher.patterns == [b"\xff\xfe", b"\x00\x01"]
    assert matcher.find_all(b"zz\xff\xfezz") == [(2, 0)]

    restored = pickle.loads(pickle.dumps(matcher))
    assert restored.patterns == matcher.patterns


def test_pickle_carries_patterns_not_the_trie():
    # The automaton is rebuilt on load: deterministic, and far simpler than
    # serialising failure links. Cheap because construction is the only cost.
    matcher = PatternMatcher(["he", "she", "hers"])

    blob = pickle.dumps(matcher)

    assert b"hers" in blob

File: tests/test_matcher.cpp
// Unit tests for the algorithm, with no Python anywhere in sight.
//
// The division of labour with tests/test_matcher.py is deliberate. This file
// owns the behaviour the binding layer cannot reach, plus the fixed-case core
// behaviour that does not need a Python interpreter to state. The Python suite
// owns the Hypothesis property tests against a brute-force oracle - which have
// no C++ equivalent worth having - along with the str/bytes overloads, pickle,
// and packaging.
//
// Note what is *not* here: step(), child(), build_failure_links() and scan() are
// private members, and module export controls which names an importer may spell,
// not class access. Reaching them would mean a friend declaration in
// aho_corasick.cppm - the one file whose selling point is that it exposes
// nothing, and the file that rebuilds every importer when edited. Their
// correctness is pinned through the public surface by the oracle instead.

#include <catch2/catch_test_macros.hpp>

#include <algorithm>
#include <cstddef>
#include <stdexcept>
#include <string>
#include <tuple>
#include <utility>
#include <vector>

import aho_corasick;

using ac::PatternMatcher;
using Match = PatternMatcher::Match;
using Offsets = PatternMatcher::Offsets;

namespace {

// The constructor copies each pattern into the matcher, so the span does not
// need to outlive the call and a temporary vector is safe here.
PatternMatcher matcher_for(std::vector<std::string> patterns) {
    return PatternMatcher(patterns);
}

}  // namespace

// --- Behaviour the bindings cannot reach -------------------------------------
//
// bindings.cpp hardwires the offset unit to the input type: bytes -> Bytes,
// str -> Characters. The two cross combinations are unreachable from Python, as
// is Pattern::chars, which the bindings drop entirely.

TEST_CASE("byte offsets are available for text that is not plain ASCII") {
    // From Python this pairing requires bytes input. In C++ the unit is a
    // parameter, so multibyte text can be scanned with byte offsets directly.
    const auto matcher = matcher_for({"bar"});
    const std::string text = "caf\xc3\xa9 bar";  // "café bar"

    const auto found = matcher.find_all(text, Offsets::Bytes);

    REQUIRE(found.size() == 1);
    CHECK(found[0] == Match{6, 0});
    CHECK(text.substr(6, 3) == "bar");

    // The same text through the other unit, which is what Python's str path
    // would report: "bar" is at character 5 but byte 6.
    const auto by_char = matcher.find_all(text, Offsets::Characters);
    REQUIRE(by_char.size() == 1);
    CHECK(by_char[0] == Match{5, 0});
}

TEST_CASE("character offsets over invalid UTF-8 are meaningless but well defined") {
    // aho_corasick.cppm claims Offsets::Characters "assumes valid UTF-8; on
    // arbitrary bytes the character positions are meaningless, but harmless,
    // with no undefined behaviour". Nothing could test that from Python, because
    // arbitrary bytes always take the Bytes path there.
    const auto matcher = matcher_for({"zz"});
    const std::string text = std::string("\xff\xfe") + "zz" + "\xc0" + "zz";

    const auto by_bytes = matcher.find_all(text, Offsets::Bytes);
    const auto by_chars = matcher.find_all(text, Offsets::Characters);

    // Same scan, so the same occurrences in the same order - only the unit the
    // offsets are reported in differs.
    REQUIRE(by_bytes.size() == 2);
    REQUIRE(by_chars.size() == by_bytes.size());

    for (std::size_t i = 0; i < by_bytes.size(); ++i) {
        CHECK(by_chars[i].pattern_index == by_bytes[i].pattern_index);

        // A character offset counts UTF-8 lead bytes, which is at most one per
        // byte, so it can never exceed the byte offset however invalid the
        // input is. It also cannot underflow: the matched bytes are themselves
        // part of the text, so the lead-byte count up to the end of a match is
        // always at least the pattern's own.
        CHECK(by_chars[i].start <= by_bytes[i].start);
        CHECK(by_chars[i].start < text.size());
    }
}

TEST_CASE("a pattern records its length in code points as well as bytes") {
    // The bindings expose patterns as bytes and drop Pattern::chars, so this
    // field has no Python-visible effect at all - but find_all's character
    // offsets are computed from it.
    const auto matcher = matcher_for({"caf\xc3\xa9", "ab"});

    const auto patterns = matcher.patterns();

    REQUIRE(patterns.size() == 2);
    CHECK(patterns[0].text.size() == 5);  // "café" is five bytes
    CHECK(patterns[0].chars == 4);        // and four code points
    CHECK(patterns[1].text.size() == 2);  // ASCII: the two agree
    CHECK(patterns[1].chars == 2);
}

TEST_CASE("an empty pattern is rejected as std::invalid_argument") {
    // Python only ever sees this as ValueError, via nanobind's translation. The
    // exception type itself is part of the C++ interface and is only observable
    // here.
    CHECK_THROWS_AS(matcher_for({"ok", ""}), std::invalid_argument);
}

TEST_CASE("a matcher can be copied and moved") {
    // nanobind placement-news a PatternMatcher and holds it by pointer, so these
    // operations are never exercised through the bindings.
    const auto original = matcher_for({"he", "she"});

    SECTION("a copy scans independently and leaves the original usable") {
        const PatternMatcher copy = original;

        CHECK(copy.find_all("ushers", Offsets::Bytes) == original.find_all("ushers", Offsets::Bytes));
        CHECK(copy.num_patterns() == 2);
        CHECK(original.num_patterns() == 2);
    }

    SECTION("a move carries the automaton") {
        PatternMatcher source = matcher_for({"he", "she"});
        const PatternMatcher moved = std::move(source);

        CHECK(moved.num_patterns() == 2);
        CHECK(moved.count("ushers") == 2);
    }
}

// --- Core behaviour ----------------------------------------------------------
//
// Moved here from tests/test_matcher.py: fixed cases that need no interpreter to
// state, and that fail faster and point at C++ line numbers when they break.

TEST_CASE("overlapping and nested matches are all reported") {
    // The textbook case: "he" is nested inside "she", and "hers" overlaps both.
    // A loop of find would report only some of these.
    const auto matcher = matcher_for({"he", "she", "hers"});

    auto found = matcher.find_all("ushers", Offsets::Bytes);
    std::ranges::sort(found, [](const Match& a, const Match& b) {
        return std::tie(a.start, a.pattern_index) < std::tie(b.start, b.pattern_index);
    });

    CHECK(found == std::vector<Match>{{1, 1}, {2, 0}, {2, 2}});
}

TEST_CASE("occurrences are counted with overlaps") {
    // "aaa" contains "aa" twice when overlaps count.
    CHECK(matcher_for({"aa"}).count("aaa") == 2);
}

TEST_CASE("matches stops at the first hit") {
    const auto matcher = matcher_for({"needle"});

    CHECK(matcher.matches("haystack needle haystack"));
    CHECK_FALSE(matcher.matches("haystack haystack"));
}

TEST_CASE("shared prefixes are merged into one trie") {
    const auto matcher = matcher_for({"he", "she", "hers"});

    CHECK(matcher.num_patterns() == 3);
    // root + "he" (2) + "she" (3) + "hers" reusing the "he" prefix and adding
    // "r","s" (2) = 8. Merging shared prefixes is the point of a trie.
    CHECK(matcher.num_states() == 8);
}

TEST_CASE("a matcher with no patterns matches nothing") {
    const auto matcher = matcher_for({});

    CHECK(matcher.num_patterns() == 0);
    CHECK_FALSE(matcher.matches("anything at all"));
    CHECK(matcher.find_all("anything at all", Offsets::Bytes).empty());
}

TEST_CASE("a repeated pattern is reported once per registration") {
    const auto matcher = matcher_for({"ab", "ab"});

    auto found = matcher.find_all("ab", Offsets::Bytes);
    std::ranges::sort(found, [](const Match& a, const Match& b) {
        return a.pattern_index < b.pattern_index;
    });

    CHECK(matcher.num_patterns() == 2);
    CHECK(found == std::vector<Match>{{0, 0}, {0, 1}});
}
